### ISSUE TITLE
Some files are skipped during gulp.dest after restore

### ISSUE THREAD

**Issue by vovayatsyuk** [sindresorhus/gulp-filter#29]
Many folders stays empty when using 1.0.2. ~~1.0.1 works fine.~~

My code:

```js
gulp.task('demo-sources', [/*'composer', */'demo-sources-cleanup'], function(cb) {
    var helperFilter = filter('**/Helper/Data.php');
    return gulp.src(['code/**/*'])
        .pipe(helperFilter)
        .pipe(helperFilter.restore())
        .pipe(gulp.dest('demo-sources/raw'));
});
```

Sorry that I didn't test at all, maybe it's my fault. I will test it it later.

**Comment by nevcos** [comment]
@nfroidure tested and passed on windows too :+1: :

* 31~33 files being outputed without the fix;
* 124 files being outputed with your fix.

Thank you!

**Comment by nfroidure** [comment]
Never understood why encouraging the use of through2, event-stream etc... It hides streams behavior for saving a few chars and introduce more bug risks.

**Comment by sindresorhus** [comment]
Because the built-in streams implementation is horrible, and anything there to make it easier is great.

Lets get thourgh2 fixed then. Can you open a ticket there?

**Comment by nevcos** [comment]
Have tried different through2 versions and forcing version "0.4.2" solves this problem.

Checked the differences between "0.4.2" and "0.5.0" and the problem comes with this tiny commit:
https://github.com/rvagg/through2/commit/7ce8c780b1219c1ebea7717e9dc111bbe4dc2ffa

From [nodejs steam documentation](http://nodejs.org/api/stream.html), highWaterMark is the "Number The maximum number of bytes to store in the internal buffer before ceasing to read from the underlying resource", so probably the steam becames full and some resources are lost in the way.

Tested a value of "128" and the problem did not occur (in my case). Not knowing much about streams, I think it's a bad idea limiting the value without at least giving a warning (when the level is reached).

**Comment by sindresorhus** [comment]
@nevcos Well, it was already limited to `16kb` before that change or in our use-case `32224` files. The change was actually a good thing as it brought forward a flushing bug somewhere. Most likely here.

**Comment by mafintosh** [comment]
I'm pretty sure this is the bug is the following

``` js
.pipe(helperFilter)
.pipe(helperFilter.restore())
```

This implies that when `helperFilter` ends then whatever streams is returned from `helperFilter.restore()` is also ended (as per regular pipe semantics).

However if you look at the current implementation of `helperFilter.restore()` it contains the following

``` js
return restoreStream.pipe(through.obj(), {end: false});
```

Now this means that `through.obj()` is returned from `stream.restore()` which (if you look above) means that when `helperFilter` ends `through.obj()` is ended.

The consequence of this is that there is not guarantee that the pipe `restoreStream.pipe(through.obj())` will complete (since `through.obj()` is ended prematurely). This is why you are seeing incomplete results.

Now, increasing hwm will workaround this bug since that means that fewer things run async (everything is read into memory) but that just means that we aren't really streaming.

**Comment by nfroidure** [comment]
@mafintosh nice catch!

I assume this issue could then be solved by adding something like those lines:
```js
var tmpStream = through.obj();
restoreStream.pipe(tmpStream, {end: false});
return duplexer(restoreStream, tmpStream);
```

On the other hand, the entire module logic could be changed to handle the end option in a more reliable way than using an intermediar stream.

@sindresorhus what about it?

**Comment by mafintosh** [comment]
@nfroidure yes on the top of my head that would fix it although this logic is still a bit convoluted

**Comment by sindresorhus** [comment]
> On the other hand, the entire module logic could be changed to handle the end option in a more reliable way than using an intermediar stream.

Any idea how? Would love to see a PR :)

**Comment by nfroidure** [comment]
Have one ;). I'll give it a go this week-end.

**Comment by nfroidure** [comment]
Had to fix it the ugly way since there is a misconception in the current API.

The fact you do not have to declare your intention of using the restore stream or not at the filter stream creation means you cannot solve the issue by using a Duplex half opened stream to handle the filtering logic. My attempts to do so just failed.

Indeed, if you do not use the restore stream, then, the filter stream will never end since streams backpressure mecanisms will impeach it to do so.

My point is to create a new major version with a new API where you declare the intent to use the restore stream at the filter level:
```js
var jsFilter = filter('**/Helper/Data.php', { restore: true });

gulp.src('**/*')
  .pipe(jsFilter())
  .pipe(doSomething())
  .pipe(jsFilter.restoreStream)
  .pipe(gulp.dest('.');
```

It will mean better performances/memory usage for simple filtering and a better implementation of the restore stream.

**Comment by sindresorhus** [comment]
The issue is fixed in 2.0.1.

@nfroidure Yeah, let's go with that. I'd like to have an optimal API for this.

**Comment by nfroidure** [comment]
@sindresorhus i find out that the gulp-filter idea was in fact really nice but i didn't found any generic library to do the same thing for raw Node stream.

It was a good occasion to make a proof of concept of the above API https://github.com/nfroidure/streamfilter

I find out that e restore option wasn't enough. In fact, coupling her with a passthough one, still at the filter instanciation level, make more sense and largely simplify the implementation.

So, if you're ok with this API:

```js

var jsFilter = filter('**/Helper/Data.php', { restore: true, passthrough: true });
// passthrough could imply restore, i'm not a big fan of implicit options but would make it shorter

gulp.src('**/*')
  .pipe(jsFilter)
  .pipe(doSomething())
  .pipe(jsFilter.restoreStream)
  .pipe(gulp.dest('.'));
```
I'm ok to start a 3.0.0 version based on those principles.

If any of you have an idea for a better API, please let me know.

**Comment by sindresorhus** [comment]
When would not want it to be `passthrough`? And does `passthrough` apply to the stream, restore stream or both?

**Comment by nfroidure** [comment]
Restore stream not beeing a PassThrough is simply the transposition of the restore() end option (see https://github.com/sindresorhus/gulp-filter/issues/12#issuecomment-39631536 ).

But you point out the fact that the restore stream is a passtrough per default in the current gulp-filter API. Since it is the most frequent usage, we could have passthrough defaulting to true.

Passthrough will apply to the restore stream only since the filter stream need to be a PassTrough stream to work. But i think i got your point, it could confuse users.

**Comment by sindresorhus** [comment]
Happy about the other stuff though.

**Comment by pikeas** [comment]
Just read through this whole thread only to find out that the issue was patched. Since this is fixed, I suggest this issue be closed and a new one opened for the proposed 3.0 changes.

**Comment by nfroidure** [comment]
Let's close this and continue the discussion here https://github.com/sindresorhus/gulp-filter/pull/39
